src/internal_rep.rs

```diff
@@ -1539,6 +1552,311 @@ fn call_to_fc_rules<'a>(
     Ok(ret)
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum FSContextType {
+    XAttr,
+    Task,
+    Trans,
+    GenFSCon,
+}
+
+impl fmt::Display for FSContextType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                FSContextType::XAttr => "xattr",
+                FSContextType::Task => "task",
+                FSContextType::Trans => "trans",
+                FSContextType::GenFSCon => "genfscon",
+            }
+        )
+    }
+}
+
+impl FromStr for FSContextType {
+    type Err = ();
+    fn from_str(s: &str) -> Result<FSContextType, ()> {
+        match s {
+            "xattr" => Ok(FSContextType::XAttr),
+            "task" => Ok(FSContextType::Task),
+            "trans" => Ok(FSContextType::Trans),
+            "genfscon" => Ok(FSContextType::GenFSCon),
+            _ => Err(()),
+        }
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Clone, Debug, PartialEq, PartialOrd, Ord)]
+pub struct FileSystemContextRule<'a> {
```

Comment:
We should generally be storing CascadeStrings here so that we don't need to go back to the FuncCall later to get ranges.